pkg/cover/backend/mach-o.go

```diff
@@ -0,0 +1,146 @@
+		&containerFns{
+			readSymbols:           machoReadSymbols,
+			readCoverPoints:       machoReadCoverPoints,
+			readModuleCoverPoints: machoReadModuleCoverPoints,
+			readTextRanges:        machoReadTextRanges,
+		},
+	)
+}
+
+func machoReadSymbols(module *Module) ([]*Symbol, *symbolInfo, error) {
+	file, err := macho.Open(module.Path)
+	if err != nil {
+		return nil, nil, err
+	}
+	text := file.Section("__text")
+	if text == nil {
+		return nil, nil, fmt.Errorf("no __text section in the object file")
+	}
+	if file.Symtab == nil {
+		return nil, nil, fmt.Errorf("failed to read Mach-O symbols")
+	}
+	var symbols []*Symbol
+	info := &symbolInfo{
+		textAddr:    text.Addr,
+		traceCmp:    make(map[uint64]bool),
+		tracePCIdx:  make(map[int]bool),
+		traceCmpIdx: make(map[int]bool),
+	}
+
+	// We don't get symbol lengths or symbol ends in Mach-O symbols. So we
+	// guesstimate them by taking the next symbols beginning -1. That only
+	// works after we have sorted them.
+	sort.Slice(file.Symtab.Syms, func(i, j int) bool {
+		return file.Symtab.Syms[i].Value < file.Symtab.Syms[j].Value
+	})
+
+	for i, symb := range file.Symtab.Syms {
+		// Mach-Os doesn't contain the Symbol size like in ELF
+		symbEnd := text.Addr + text.Size
+		if i < len(file.Symtab.Syms)-1 {
+			symbEnd = file.Symtab.Syms[i+1].Value - 1
+		}
+
+		text := symb.Value >= text.Addr && symbEnd <= text.Addr+text.Size
+		if text {
+			symbStart := symb.Value + module.Addr
+			symbols = append(symbols, &Symbol{
+				Module: module,
+				ObjectUnit: ObjectUnit{
+					Name: symb.Name,
+				},
+				Start: symbStart,
+				End:   symbEnd,
+			})
+		}
+		if strings.HasPrefix(symb.Name, "___sanitizer_cov_trace_") {
+			if symb.Name == "___sanitizer_cov_trace_pc_guard" {
+				info.tracePCIdx[i] = true
+				if text {
+					info.tracePC = symb.Value
+				}
+			} else {
+				info.traceCmpIdx[i] = true
+				if text {
+					info.traceCmp[symb.Value] = true
+				}
+			}
+		}
+	}
+	return symbols, info, nil
+}
+
+func machoReadTextRanges(module *Module) ([]pcRange, []*CompileUnit, error) {
+	dir, kernel := filepath.Split(module.Path)
+	dSYMPath := filepath.Join(dir, fmt.Sprintf(
+		"%[1]s.dSYM/Contents/Resources/DWARF/%[1]s", kernel))
+	dSYM, err := macho.Open(dSYMPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	debugInfo, err := dSYM.DWARF()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse DWARF: %v", err)
+	}
+	return readTextRanges(debugInfo, module, nil)
+}
+
+func machoReadCoverPoints(target *targets.Target, module *Module,
+	info *symbolInfo) ([2][]uint64, error) {
+	file, err := macho.Open(module.Path)
+	if err != nil {
+		return [2][]uint64{}, err
+	}
+	text := file.Section("__text")
+	if text == nil {
+		return [2][]uint64{}, fmt.Errorf("no __text section in the object file")
+	}
+	data, err := text.Data()
+	if err != nil {
+		return [2][]uint64{}, fmt.Errorf("failed to read __text: %v", err)
+	}
+
+	return readCoverPoints(target, info, &textWrap{
+		addr: text.Addr,
```

Comment:
info already contains text.Addr. Can we remove textWrap and pass in just data?

Reply:
Ah yes, much better.